File: app/src/main/java/net/emilla/struct/sort/Searchable.java
package net.emilla.struct.sort;

import net.emilla.util.Strings;

public interface Searchable<T extends Searchable<T>> extends Comparable<T> {

    String ordinal();

    @Override
    default int compareTo(T other) {
        return ordinal().compareToIgnoreCase(other.ordinal());
    }

    default boolean ordinalIs(String search) {
        return ordinal().equalsIgnoreCase(search);
    }

    default boolean ordinalContains(String search) {
        return Strings.containsIgnoreCase(ordinal(), search);
    }

}

File: app/src/main/java/net/emilla/struct/sort/Searchable.kt
package net.emilla.struct.sort

import net.emilla.util.containsIgnoreCase

abstract class Searchable<T : Searchable<T>> : Comparable<T> {
    abstract fun ordinal(): String

    override fun compareTo(other: T) = compareTo(other.ordinal())
    fun compareTo(search: String) = ordinal().compareTo(search, ignoreCase = true)

    fun ordinalIs(search: String) = ordinal().equals(search, ignoreCase = true)
    fun ordinalContains(search: String) = ordinal().containsIgnoreCase(search)
}
